### ISSUE TITLE
feat(ngMessages): new ng-messages-options to support sane defaults

### ISSUE THREAD

**Issue by manikantag** [angular/angular.js#10847]
Hi,

As it is very common to show the ng-messages only when form is submitted or when user interacts with the field. Now we do this using something similar to below:

	<div ng-messages="loginForm.userId.$error" ng-if="loginForm.$submitted || loginForm.userId.$touched">
		<div ng-message="required">User ID can't be empty</div>
		<div ng-message="minlength">User ID must be at least 6 chars long</div>
	</div>

As in most cases, `ng-if="loginForm.$submitted || loginForm.userId.$touched"` is used, we can extend the ngMessages functionality by using the same technique of `ng-model-options`, `ng-messages-options` can be implemented to support this common behaviour. 

Ex:

	<div ng-messages="loginForm.userId.$error" ng-messages-options="{show: {touched: true, init: false}}">
		<div ng-message="required">User ID can't be empty</div>
		<div ng-message="minlength">User ID must be at least 6 chars long</div>
	</div>

or, something similar to this *(may be even simpler!)*. I m thinking of `ng-messages-options` just to minimise yet-another attribute, like ng-messages-init="false".

**One benefit of this approach, apart from syntactic sugar, is that we can avoid the additional watcher required for `ng-if`**

**Comment by petebacondarwin** [comment]
You can do this already by adding it to the ng-messages attribute:

```html
<form name="f">
    <input ng-model="x" name="x" required ng-maxlength="10">
    {{ x }}
  </form>
  <div ng-messages="f.x.$touched && f.x.$error">
    <div ng-message="required">This is required</div>
    <div ng-message="maxlength">Too long</div>
  </div>
```

See http://plnkr.co/edit/zIJEmSPaqkj8Y720XZrC?p=preview

**Comment by petebacondarwin** [comment]
Adding an extra options attribute adds unnecessary extra code. We should document that you can control display via the `ng-messages` attribute for other users.

@manikantag - is this acceptable? Would you like to provide a PR to document the way suggested above?

**Comment by manikantag** [comment]
Clever use of logical operators... We do this all the time, but sometimes it just doesn't strike!

@petebacondarwin Sure. This technique can potentially increase the performance in validation rich forms, as we will be avoiding one more additional $watch required for `ng-if`. That is one watch for every element! For this very reason, ngMessages is recommended for these purpose over `ng-show`/`ng-hide`. This can also be documented and newbies might find it best practice.

*For JS newbies, this why it works:* 
When Angular evaluates the expression `f.x.$dirty && f.x.$error`, when `f.x.$dirty` (or, $touched, $pristine, ...) evaluates to `true`, then the final outcome of the entire expression would be `f.x.$error`, which is what exactly ngMessages need. 

Simply put, `console.log(true && {val: 2})` --> `{val: 2}`

Other way is not possible, though - ie, `f.x.$error && f.x.$dirty` won't work, as the expression only returns either `true` or `false`. But ngMessages needs an object of (some) keys with boolean values.
